src/components/profile/EnvironmentForm.jsx

```diff
@@ -0,0 +1,218 @@
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+import { withStyles } from "@material-ui/core";
+import { compose } from 'redux';
+import { connect } from "react-redux";
+import { getEnvironments } from "../../actions/environments";
+import DialogTitle from '@material-ui/core/DialogTitle';
+import Dialog from '@material-ui/core/Dialog';
+import TextField from '@material-ui/core/TextField';
+import Button from '@material-ui/core/Button';
+import api from '../../utils/api';
+import AddIcon from '@material-ui/icons/Add';
+import IconButton from '@material-ui/core/IconButton';
+import Input from '@material-ui/core/Input';
+import InputLabel from '@material-ui/core/InputLabel';
+import InputAdornment from '@material-ui/core/InputAdornment';
+import FormControl from '@material-ui/core/FormControl';
+import Chip from '@material-ui/core/Chip';
+import List from '@material-ui/core/List';
+import Typography from "@material-ui/core/Typography";
+
+
+const styles = theme => ({
+	root:{
+		flex: 1,
+	},
+	container: {
+		display: 'flex',
+		flexWrap: 'wrap',
+	},
+	textField: {
+		margin: theme.spacing.unit*2,
+	},
+	tagStyle: {
+		margin: theme.spacing.unit,
+	},
+})
+
+class EnvironmentForm extends Component {
+  constructor(props) {
+		super(props);
+		this.state = {
+			id: this.props.envExists ? this.props.environment.id : this.props.repository.id,
+			name: this.props.envExists ? this.props.environment.name : this.props.repository.name,
+			description: this.props.envExists ? this.props.environment.description : this.props.repository.description,
+			tag: '',
+			tags: this.props.envExists && Boolean(this.props.environment.tags) ? this.props.environment.tags : [],
+			error: '',
+		}
+    this.getEnvironments = this.getEnvironments.bind(this);
+  };
+
+
+	getEnvironments() {
+		this.props.getEnvironments();
+	};
+
+	handleClose = () => {
+    this.props.onClose();
+	};
+
+	handleSubmit = (event) => {
+		event.preventDefault();
+		if (this.props.envExists) {
+			api.editEnvironment(this.state).then(this.handleSuccess).catch(this.handleFailure);
+		} else {
+			const { name, description, id, tags} = this.state;
+			api.createEnvironment(name, description, id, tags).then(this.handleSuccess).catch(this.handleFailure);
+		}
+	}
+
+	handleSuccess = () => {
+		this.handleClose();
+    	api.environments().then((json) => {
+	  		this.props.getEnvironments(json.data);
+		});
+	}
+	
+	handleFailure = () => {
+		this.setState({ error: 'Request did NOT succeed!' })
+	}
+	
+
+	handleChange = name => event => {
+		this.setState({
```

Comment:
just a note, I like to add `event.preventDefault()` in these. There is a type of bug where say, on an input field, you do something like this accidentally:
```
<input onChange={this.handleChange('name')} />
```
But you forget to add the `value` property. This will still appear on the UI correctly because the native HTML element will still fill in the keystrokes into the field. But it's possible that what's in your `this.state.name` is not the same value (if you're doing any processing, debouncing, etc.).  

`event.preventDefault()` makes it so the keystroke will not be set directly in the HTML element, so if you type but forget the value it will just appear as if nothing is happening (which is really what we would expect in this type of bug).

Reply:
I tried to make that error happen. However, if I write 

```
<TextField
    value={this.state.name}
    onChange={this.handleChange()} 
/>
```

instead, nothing bad really happens. I can't even make errors work properly...